Refactor GravatarServices to expose suspend functions

hamorillo opened the pull request:
Closes #144 

### Description

This PR tries to improve our approach with Kotlin Coroutines. As we discussed in #143 ([here](https://github.com/Automattic/Gravatar-SDK-Android/pull/143#discussion_r1588417398)), we should use retrofit with its coroutines compatible API. 

We've modified the OpenApi generator config to use this when generating the `Service` classes. Now, they will be `suspend` functions returning `Response<*>`. That way, the caller can handle the `CoroutineScope`.

### Testing Steps

- Run the demo-app
- Try to retrieve different profiles
- Try to upload a new avatar _(You'll need your Access Token)_

hamorillo committed:
Refactor GravatarServices to expose suspend functions

hamorillo commented:
@maxme I haven't finished this PR yet. I need to do a second pass about the code, but it would be nice to have your opinion before your AFK. (If possible, I know there's only one day left, so if you can't, no worries at all.)

hamorillo commented on gravatar/src/main/java/com/gravatar/di/container/GravatarSdkContainer.kt:
```diff
@@ -10,9 +10,15 @@ import okhttp3.OkHttpClient
+/**
+ * Container that handle the DI for the Gravatar SDK.
+ */
+public class GravatarSdkContainer private constructor() {
```
My main concert is about making this class public so that it can be used in the `gravatar-ui ` module. I wanted to use the dispatcher in the [GravatarImagePickerWrapper.kt](https://github.com/Automattic/Gravatar-SDK-Android/pull/151/files#diff-30259c0c1479d08a6f6951482b4f25dee6b030d4e33db9ea7ccc01ffefbbe529), but maybe I should look for another option. Wdty?

Note that not everything will be public, but still not sure if that's the best way.

maxme replied:
What are our options?
- what you've done: make this class and fields public.
- have another internal container for `gravatar-ui`. We'll need to duplicate a few things (dispatchers for now, maybe more later). This would make sense if someone could use `gravatar-ui` without `gravatar`, but I don't think we'll ever do this. 

I think it's fine making these fields public, I doubt we'll need to override them in other context than prod/testing environments so even if a third party use them by mistake, it shouldn't have a big impact.

eyedol replied:
If you make the suspend functions main safe you might not need to expose it but if you want to make the dispatchers defined in it for general use, then I guess relaxing its visibility is the way to go.

maxme commented on gravatar/src/main/java/com/gravatar/api/apis/ProfilesApi.kt:
```diff
@@ -25,11 +25,11 @@ internal interface ProfilesApi {
      */
     @GET("profiles/{profileIdentifier}")
-    fun getProfileById(
+    suspend fun getProfileById(
```
Reading the auto generated documentation made me realize we don't have an error defined in`ErrorType` for 429 (Rate limit exceeded). We should add it asap (either in this PR or create a new ticket).

hamorillo replied:
[Ticket created](https://github.com/Automattic/Gravatar-SDK-Android/issues/153)

hamorillo committed:
Sorting test files in their corresponding packages

hamorillo committed:
Add unit test for `ProfileService

hamorillo commented:
The changes are ready for a first review. 

@thomashorta @eyedol, as you participated in the original discussion, feel free to express your opinions.

maxme reviewed (approved):
LGTM :shipit: 

👍 Thanks for adding new tests to the ProfileService.

Error management is the same as before (http code or exceptions are converted into an ErrorType and forwarded to the listener via the `Result` class). If you don't want to deal with the 429 error in this PR, please open a new ticket.

eyedol commented on demo-app/src/main/java/com/gravatar/demoapp/ui/DemoGravatarApp.kt:
```diff
@@ -248,7 +249,7 @@ private fun ProfileTab(modifier: Modifier = Modifier, onError: (String?, Throwab
                     onClick = {
                         keyboardController?.hide()
-                        scope.launch {
+                        scope.launch(Dispatchers.IO) {
```
If the suspend function is `main-safe` the caller won't have to provide a dispatcher for it. Similar to how Retrofit callback calls work. In general better to design the `suspend` function to define the dispatcher it should run in

eyedol commented on gravatar-ui/src/main/java/com/gravatar/ui/GravatarImagePickerWrapper.kt:
```diff
@@ -49,11 +54,20 @@ public fun GravatarImagePickerWrapper(
             UCrop.getOutput(intentData)?.let { croppedImageUri ->
                 listener.onAvatarUploadStarted()
-                AvatarService().upload(croppedImageUri.toFile(), Email(email), wordpressBearerToken, listener)
+                coroutineScope.launch(GravatarSdkDI.dispatcherIO) {
```
Same comment as earlier

eyedol commented on gravatar/src/main/java/com/gravatar/GravatarApiService.kt:
```diff
@@ -12,9 +12,9 @@ import retrofit2.http.Part
     @Multipart
     @POST("upload-image")
-    fun uploadImage(
+    suspend fun uploadImage(
```
👍🏿

eyedol commented on gravatar/src/main/java/com/gravatar/services/AvatarService.kt:
```diff
@@ -22,51 +16,35 @@ public class AvatarService(private val okHttpClient: OkHttpClient? = null) {
-        gravatarUploadListener: GravatarListener<Unit, ErrorType>,
-    ) {
+    public suspend fun upload(file: File, email: Email, wordpressBearerToken: String): Result<Unit, ErrorType> {
         val service = GravatarSdkDI.getGravatarApiV1Service(okHttpClient)
```
I know this is not part of your changes but I think it's best if this is defined as a class property instead of a local val as it seems for every time the `upload` function is called, it  setups the underlying retrofit service. I guess we could avoid doing that as I don't see anything changing upon call of the `upload` function that would require recreating the retrofit service. But yeah I stand corrected if I'm missing something and hence it was designed this way.

hamorillo replied:
Nice suggestion! 🚀

eyedol commented on gravatar/src/main/java/com/gravatar/services/AvatarService.kt:
```diff
@@ -22,51 +16,35 @@ public class AvatarService(private val okHttpClient: OkHttpClient? = null) {
-        wordpressBearerToken: String,
-        gravatarUploadListener: GravatarListener<Unit, ErrorType>,
-    ) {
+    public suspend fun upload(file: File, email: Email, wordpressBearerToken: String): Result<Unit, ErrorType> {
```
I would define this main safe as recommended in the [coroutines best practices doc](https://developer.android.com/kotlin/coroutines/coroutines-best-practices#main-safe)

eyedol commented on gravatar/src/main/java/com/gravatar/services/AvatarService.kt:
```diff
@@ -22,51 +16,35 @@ public class AvatarService(private val okHttpClient: OkHttpClient? = null) {
+                )
+                Result.Failure(errorTypeFromHttpCode(response.code()))
+            }
+        } catch (ex: Exception) {
+            Result.Failure(ex.error())
+        }
```
It seems you are catching all exceptions which isn't safe as you're also catching `CancellationException`. Conventional advice is you re-throw cancellation exceptions.

```kotlin
...
return try {
    val response = service.uploadImage("Bearer $wordpressBearerToken", identity, filePart)

    if (response.isSuccessful) {
        Result.Success(Unit)
    } else {
        // Log the response body for debugging purposes if the response is not successful
        Logger.w(
            LOG_TAG,
            "Network call unsuccessful trying to upload Gravatar: $response.body",
        )
        Result.Failure(errorTypeFromHttpCode(response.code()))
    }
} catch( ce: CancelationException) {
    throw ce
} catch (ex: Throwable) {
    Result.Failure(ex.error())
}

```

eyedol commented on gravatar/src/main/java/com/gravatar/services/ProfileService.kt:
```diff
@@ -22,95 +15,34 @@ public class ProfileService(private val okHttpClient: OkHttpClient? = null) {
+                )
+                Result.Failure(errorTypeFromHttpCode(response.code()))
+            }
+        } catch (ex: Exception) {
+            Result.Failure(ex.error())
```
Same as comment above, re-throw the `CancelationException`

eyedol commented on gravatar/src/main/java/com/gravatar/services/ProfileService.kt:
```diff
@@ -22,95 +15,34 @@ public class ProfileService(private val okHttpClient: OkHttpClient? = null) {
      * @param hashOrUsername The hash or username to fetch the profile for
      * @return The fetched profile
      */
     public suspend fun fetch(hashOrUsername: String): Result<Profile, ErrorType> {
```
Same as above, make this main-safe if possible

eyedol commented on gravatar/src/main/java/com/gravatar/services/ProfileService.kt:
```diff
@@ -22,95 +15,34 @@ public class ProfileService(private val okHttpClient: OkHttpClient? = null) {
-                    }
-                },
-            )
+        val service = GravatarSdkDI.getGravatarApiV3Service(okHttpClient)
```
Also this `service` local variable can be defined as a property to avoid recreating the retrofit service like commented above

eyedol reviewed (commented):
Nice seeing these changes coming to live. In general it looks good. Passed a few comments worth looking into. Thanks for picking this up so fast.

hamorillo committed:
Make main-safe the methods in Avatar and Profile services

If we make the methods in services main-safe, we don't need to expose dispatchers in GravatarSdkContainer, and callers don't need to care where they execute the call.

Now, network calls are handled by the SDK and will always be executed in the IO dispatcher.

hamorillo committed:
Rethrow CancellationException on services

We were catching all exceptions on Services' calls.

It looks like it's not the best practice to catch CancellationException because it can cause unexpected behaviors while canceling the stack of coroutines.

hamorillo committed:
Avoid service recreation on each method invocation

hamorillo commented:
Thanks for your comments @eyedol! I've improved the code with some of your comments.

eyedol commented on gravatar/src/main/java/com/gravatar/services/AvatarService.kt:
```diff
@@ -2,71 +2,51 @@ package com.gravatar.services
-                        }
-                    }
-                }
+        return runCatchingService {
```
👍🏿

eyedol commented on gravatar/src/main/java/com/gravatar/services/AvatarService.kt:
```diff
@@ -2,71 +2,51 @@ package com.gravatar.services
-                    }
-                }
+        return runCatchingService {
+            withContext(Dispatchers.IO) {
```
For consistency perhaps use this one `GravatarSdkDI.dispatcherIO`? As it seems to be the one managed by us.

hamorillo replied:
Good catch! Thanks.

eyedol commented on gravatar/src/main/java/com/gravatar/services/ServiceUtils.kt:
```diff
@@ -0,0 +1,14 @@
+
+import kotlinx.coroutines.CancellationException
+
+internal inline fun <T> runCatchingService(block: () -> Result<T, ErrorType>): Result<T, ErrorType> {
```
👍🏿

eyedol commented on gravatar/src/test/java/com/gravatar/services/ProfileServiceTests.kt:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    @Test
+    fun `given an username when loading its profile but date is NOT returned then result is UNKNOWN failure`() =
```
Small typo ~date~ `data`

eyedol commented on gravatar/src/test/java/com/gravatar/services/ProfileServiceTests.kt:
```diff
@@ -0,0 +1,121 @@
+            val loadProfileResponse = profileService.fetchByUsername(username)
+
+            coVerify(exactly = 1) { containerRule.gravatarApiServiceMock.getProfileById(username) }
+            assertTrue((loadProfileResponse as Result.Failure).error == ErrorType.UNKNOWN)
```
Not sure if we should test for the other error types like `TIMEOUT` similar to the test cases in `AvatarServiceTest`

hamorillo replied:
I'll probably refactor that in another PR.

eyedol reviewed (approved):
Thanks for making the changes. It looks good. Passed a few minor comments worth checking other than that LGTM.
